### PR TITLE
Add example for returning vectors

### PR DESCRIPTION
**Pull request by worr**
I left out ruby, haskell and csharp since I don't know them well enough to create worthwhile examples

### CODE REVIEW

**Comment by shepmaster** [comment]
Thanks! I swear I'll try to review this soon, but life is hectic with [Rust Belt Rust](http://www.rust-belt-rust.com/) an a little over a week!

File: examples/Makefile [hunk 1]
```diff
@@ -33,6 +33,8 @@ ifeq (${OS},Darwin)
 shared_lib = lib${1}.dylib
 else ifeq (${OS},Linux)
 shared_lib = lib${1}.so
+else ifeq (${OS},OpenBSD)
```
Comment (shepmaster) [inline examples/Makefile]: I love that you are adding FreeBSD support, but I'd rather that be split out into a separate PR.

File: examples/vector_return/Cargo.lock [hunk 2]
```diff
@@ -0,0 +1,14 @@
+[root]
+name = "vector_return"
+version = "0.1.0"
+dependencies = [
+ "libc 0.2.16 (registry+https://github.com/rust-lang/crates.io-index)",
+]
+
+[[package]]
+name = "libc"
+version = "0.2.16"
```
Comment (shepmaster) [inline examples/vector_return/Cargo.lock]: I'm pulling the rug out from under you a little bit by doing my own updates to 0.2.17 (#44). It'd be super great if you could match.

File: examples/vector_return/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,30 @@
+extern crate libc;
+
+use libc::size_t;
+use std::mem;
+
+#[no_mangle]
+pub extern fn counter_generate(start: size_t, size: size_t, vec: *mut *mut size_t) -> size_t {
```
Comment (shepmaster) [inline examples/vector_return/src/lib.rs]: I'd rather a type other than `size_t` for the element since it's easy to lose it compared to the `size` parameter.

File: examples/vector_return/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+    let mut counted = Vec::new();
+
+    for x in start..start + size as size_t {
+        counted.push(x);
+    }
```
Comment (shepmaster) [inline examples/vector_return/src/lib.rs]: Ah, I see why you used `size_t` now.

I'd prefer this to be written as

```rust
    let counted: Vec<_> = (start..).take(s).collect();
```

This has two main benefits:

1. Removes the mutability requirement.
1. Allocates the proper amount of memory for the vector all at once at the beginning.

File: examples/vector_return/src/main.c [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+
+extern void
+counter_free(size_t *vec, size_t size);
+
+int main(void) {
+  size_t *vec;
+
+  counter_generate(0, 10, &vec);
+  for (int i = 0; i < 10; i++) {
+    printf("%d..", i);
```
Comment (shepmaster) [inline examples/vector_return/src/main.c]: Is it valid to use `%d` to print a `size_t`?

File: examples/vector_return/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+}
+
+#[no_mangle]
+pub extern fn counter_free(arr: *mut size_t, size: size_t) {
+    unsafe {
+        if arr.is_null() {
+            return;
+        }
+
+        Vec::from_raw_parts(arr, size, size)
```
Comment (shepmaster) [inline examples/vector_return/src/lib.rs]: This isn't quite right. There's no guarantee that the `size` parameter and the `capacity` have to match. In fact, it's quite likely that they *don't* due to the doubling nature. One of two things needs to happen:

1. The capacity *also* needs to be returned.
1. Call [`shrink_to_fit`](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.shrink_to_fit) before decomposing the vector.

File: examples/vector_return/src/main.c [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+extern size_t
+counter_generate(size_t start, size_t size, size_t **vec);
+
+extern void
+counter_free(size_t *vec, size_t size);
+
+int main(void) {
+  size_t *vec;
+
+  counter_generate(0, 10, &vec);
```
Comment (shepmaster) [inline examples/vector_return/src/main.c]: The return value is being returned. This needs to match up with the future call to `free`.

File: examples/vector_return/src/main.js [hunk 8]
```diff
@@ -0,0 +1,25 @@
+const ffi = require('ffi');
+const ref = reqyire('ref');
```
Comment (shepmaster) [inline examples/vector_return/src/main.js]: `reqyire` doesn't seem to be valid.

File: examples/vector_return/src/main.js [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+
+const sizeArray = ArrayType(ref.types.size_t);
+const sizePtr = ref.refType(sizeArray);
+
+const lib = ffi.Library('libvector_return', {
+    counter_generate: ['size_t', ['size_t', 'size_t', sizePtr]],
+    counter_free: ['void', [sizeArray, 'size_t']],
+});
+
+function themeSongGenerate(len) {
```
Comment (shepmaster) [inline examples/vector_return/src/main.js]: It's not much of a theme song :-) A more matching name would be better

File: examples/vector_return/src/main.py [hunk 10]
```diff
@@ -0,0 +1,23 @@
+#!/usr/bin/env python2
```
Comment (shepmaster) [inline examples/vector_return/src/main.py]: Thanks for adding this, as it made me realize I had a bunch of old junk! #45 updates this to `python3` and adds a space; I'd appreciate it if you could match.

File: examples/vector_return/src/main.py [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+import sys, ctypes
+from ctypes import pointer, POINTER, c_size_t
+
+prefix = {'win32': ''}.get(sys.platform, 'lib')
+extension = {'darwin': '.dylib', 'win32': '.dll'}.get(sys.platform, '.so')
+lib = ctypes.cdll.LoadLibrary(prefix + "vector_return" + extension)
+
+lib.counter_generate.argtypes = (c_size_t, c_size_t, POINTER(POINTER(c_size_t)))
+lib.counter_generate.restype = ctypes.c_size_t
+lib.counter_free.argtypes = ((POINTER(c_size_t)), c_size_t)
```
Comment (shepmaster) [inline examples/vector_return/src/main.py]: Why are the extra parenthesis needed?

**Comment by shepmaster** [comment]
There are also JS failures, once some typos are cleared up:

```
/Users/shep/Projects/rust-ffi-omnibus/examples/node_modules/ffi/lib/_foreign_function.js:55
      throw e
      ^

TypeError: error setting argument 2 - writePointer: Buffer instance expected as third argument
    at Object.writePointer (/Users/shep/Projects/rust-ffi-omnibus/examples/node_modules/ref/lib/ref.js:742:11)
    at Object.set (/Users/shep/Projects/rust-ffi-omnibus/examples/node_modules/ref/lib/ref.js:484:13)
    at Object.alloc (/Users/shep/Projects/rust-ffi-omnibus/examples/node_modules/ref/lib/ref.js:516:13)
    at Object.proxy [as counter_generate] (/Users/shep/Projects/rust-ffi-omnibus/examples/node_modules/ffi/lib/_foreign_function.js:50:22)
    at themeSongGenerate (/Users/shep/Projects/rust-ffi-omnibus/examples/vector_return/src/main.js:15:26)
    at Object.<anonymous> (/Users/shep/Projects/rust-ffi-omnibus/examples/vector_return/src/main.js:25:13)
    at Module._compile (module.js:573:32)
    at Object.Module._extensions..js (module.js:582:10)
    at Module.load (module.js:490:32)
    at tryModuleLoad (module.js:449:12)
```

File: examples/vector_return/src/lib.rs (hunk 3 above)
Comment (shepmaster) [inline examples/vector_return/src/lib.rs]: Maybe a 16-bit number to spice things up?

File: examples/vector_return/src/main.c (hunk 5 above)
Comment (worr) [inline examples/vector_return/src/main.c]: The variable here is a `int`, so `%d` is fine.

File: examples/vector_return/src/lib.rs (hunk 6 above)
Comment (worr) [inline examples/vector_return/src/lib.rs]: I'm opting for `shrink_to_fit`, if only because having both a `cap` and `size` parameter to handle lends itself to being error-prone imo.

Thanks!!

File: examples/Makefile (hunk 1 above)
Comment (worr) [inline examples/Makefile]: OpenBSD ;)

I'll pop that into a separate commit + PR. I don't have a Linux or Mac box to test on, so it's more or less a requirement for me to get going.

File: examples/vector_return/src/main.py (hunk 11 above)
Comment (worr) [inline examples/vector_return/src/main.py]: These are tuples, so I'm being explicit about the declaration. I haven't see much python where the parens around a tuple are omitted, so I didn't even think to omit them here.

If you're against this style, I can omit them.

**Comment by worr** [comment]
I've updated the PR with the requested changes.

I can't debug the js errors on this system atm, so beyond correcting the obvious typos, I've just submitted what I have. Sorry about how awful the JS is.

File: examples/vector_return/src/main.py (hunk 11 above)
Comment (shepmaster) [inline examples/vector_return/src/main.py]: Ah, no that's fine, although I'm used to `(thing, )` with the trailing comma.

**Comment by shepmaster** [comment]
Closed via tweaked example in #48

**Comment by shepmaster** [comment]
Thank you very much! I'll work on adding the prose and the other languages sometime Real Soon Now! :-)
